docs/architecture/adr-013-proto-buf-serde.md

```diff
@@ -0,0 +1,136 @@
+Instances of [protobuf well-known types][wkt] are represented with custom types
+defined in the crate, due to historical problems with `prost-types`.
+
+Additional trait implementations are derived for some of the generated
+message structures, mainly to provide JSON serialization for RPC and possibly
+other non-Protobuf uses. This is done with custom attributes added through code
+generator options.
+While separating the unrelated concerns by e.g. implementing the traits directly on
+the domain types in the `cometbft` crate would be cleaner architecturally, doing so
+on the protoc-generated types is convenient because the protobuf and RPC
+implementations in the CometBFT Go codebase reuse the same data structures and therefore,
+the serialization schema.
+
+Rust projects which reuse protocol types from CometBFT and generate code from
+proto files that include `tendermint.*` or `cometbft.*` proto packages as a
+dependency currently use the same approach, with some local arrangements about
+mapping the WKTs and partial reuse of the serde implementations.
+This results in [problems][ibc-proto-rs#194] in dependency management that
+are easy to overlook, and duplication of types representing the protobuf WKTs.
+
+[cometbft-git]: https://github.com/cometbft/cometbft
+[wkt]: https://protobuf.dev/reference/protobuf/google.protobuf/
+[ibc-proto-rs#194]: https://github.com/cosmos/ibc-proto-rs/issues/194
+
+## Decision
+
+Switch to [buf](https://buf.build/) for generating `cometbft-proto` Rust
+bindings for the CometBFT protos that are published to the Buf Schema Registry
+under the [`cometbft/cometbft`](https://buf.build/cometbft/cometbft) module.
+
+Attempt to make use of the following plugins:
+
+* [protoc-gen-prost] to generate the core data structures.
+* [protoc-gen-prost-serde] to generate the bulk of the serde implementations
+  for the same data structures.
+* [protoc-gen-tonic] for gRPC services.
+
+[protoc-gen-prost]: https://github.com/neoeinstein/protoc-gen-prost/blob/main/protoc-gen-prost/README.md
+[protoc-gen-prost-serde]: https://github.com/neoeinstein/protoc-gen-prost/blob/main/protoc-gen-prost-serde/README.md
+[protoc-gen-tonic]: https://github.com/neoeinstein/protoc-gen-prost/blob/main/protoc-gen-tonic/README.md
+
+If the plugins are found lacking necessary features, try an [alternative approach][alt-gen]
+with `buf build --as-file-descriptor-set` and feeding the output to`tonic-build`. The serde
+implementations would need to be custom-derived like with the current approach.
+
+[alt-gen]: https://github.com/cometbft/cometbft-rs/issues/3#issuecomment-1916766090
+
+### API changes for protobuf/gRPC users
+
+The prost and tonic plugins invoke `prost-build` and `tonic-build`, respectively,
+so the code produced by them should not principally differ from the output of the
+current proto-compiler tool.
+
+### Changes in serialization
+
+The prost-serde plugin produces code depending on `pbjson`. This is
```

Comment:
Note that ibc-rs has a strict requirement for all proto types it transitively depends on to be `no_std` compatible. Unfortunately, at the moment the `pbjson` crate is not `no_std` compatible, and we've thus been using [our own `no_std` compatible fork of `pbjson`](https://github.com/informalsystems/pbjson).

The way I see it currently, we could either:

a) try and upstream our changes to `influxdata/pbjson` and update the `prost-serde` plugin to the latest version of `pbjson` once they release our changes.
b) Fork `prost-serde` to use `informalsystems/pbjson` and use that instead.

Reply:
Thanks, I support a) as a long-term solution and b) as the second best.
Is it feasible to submit the `no_std` changes in a PR to `pbjson`?

Reply:
Ok, so let's start with (b) and work towards (a) in the meantime.

> Is it feasible to submit the `no_std` changes in a PR to `pbjson`?

Yeah I can open an issue and prepare a PR. I am quite busy this week, so not sure when I'll get to it but will keep you posted.